lttoolbox/input_file.cc

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2021 Apertium
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <lttoolbox/input_file.h>
+#include <utf8.h>
+#include <stdexcept>
+#include <unicode/ustdio.h>
+#include <cstring>
+
+InputFile::InputFile()
+  : infile(stdin), buffer_size(0)
+{}
+
+InputFile::~InputFile()
+{
+  close();
+}
+
+bool
+InputFile::open(char* fname)
+{
+  close();
+  if (fname == NULL) {
+    infile = stdin;
+  } else {
+    infile = fopen(fname, "r");
+  }
+  return (infile != NULL);
+}
+
+void
+InputFile::close()
+{
+  if (infile != NULL) {
+    if (infile != stdin) {
+      fclose(infile);
+      delete infile;
```

Comment:
FILE* does not need to be deleted. That should cause a runtime error. Just set it to `nullptr`